src/IB/IBFEMethod.cpp

```diff
@@ -227,15 +231,413 @@ get_x_and_FF(libMesh::VectorValue<double>& x,
+    }
+}
 
+// for building linear system to compute DG approximation to the stress normalization function
 void
-assemble_poisson(EquationSystems& es, const std::string& /*system_name*/)
+assemble_ipdg_poisson(EquationSystems& es, const std::string& system_name)
 {
     const MeshBase& mesh = es.get_mesh();
     const BoundaryInfo& boundary_info = *mesh.boundary_info;
+    const PointLocatorTree& point_locator(mesh);
+    point_locator.build(TREE_ELEMENTS, mesh);
+
     const unsigned int dim = mesh.mesh_dimension();
     auto& system = es.get_system<LinearImplicitSystem>(IBFEMethod::PHI_SYSTEM_NAME);
+    const Real jump0_penalty = es.parameters.get<Real>("ipdg_jump0_penalty");
+    const Real jump1_penalty = es.parameters.get<Real>("ipdg_jump1_penalty");
+    const Real beta0 = es.parameters.get<Real>("ipdg_beta0");
+    const Real beta1 = es.parameters.get<Real>("ipdg_beta1");
+    const double epsilon = es.parameters.get<Real>("Phi_epsilon");
+    const double epsilon_inv = (std::abs(epsilon) > std::numeric_limits<double>::epsilon() ? 1.0 / epsilon : 0.0);
+    const Real diffusion = es.parameters.get<Real>("Phi_diffusion");
+
+    // had to remove the 'const' qualifier for dof_map because I need info about periodic boundaries
+    DofMap& dof_map = system.get_dof_map();
+    PeriodicBoundaries* periodic_boundaries = dof_map.get_periodic_boundaries();
+
+    FEType fe_type = system.variable_type(0);
+
+    UniquePtr<FEBase> fe(FEBase::build(dim, fe_type));
+    UniquePtr<FEBase> fe_elem_face(FEBase::build(dim, fe_type));
+    UniquePtr<FEBase> fe_neighbor_face(FEBase::build(dim, fe_type));
+
+    // Quadrature rules for numerical integration on volumes and faces
+    QGauss qrule(dim, fe_type.default_quadrature_order());
+    fe->attach_quadrature_rule(&qrule);
+
+    QGauss qface(dim - 1, fe_type.default_quadrature_order());
+    fe_elem_face->attach_quadrature_rule(&qface);
+    fe_neighbor_face->attach_quadrature_rule(&qface);
+
+    // for volume integrals
+    const std::vector<Real>& JxW = fe->get_JxW();
+    const std::vector<std::vector<RealGradient> >& dphi = fe->get_dphi();
+    const std::vector<std::vector<Real> >& phi = fe->get_phi();
+
+    //  for surface integrals
+    const std::vector<std::vector<Real> >& phi_face = fe_elem_face->get_phi();
+    const std::vector<std::vector<RealGradient> >& dphi_face = fe_elem_face->get_dphi();
+    const std::vector<Real>& JxW_face = fe_elem_face->get_JxW();
+    const std::vector<libMesh::Point>& qface_normals = fe_elem_face->get_normals();
+
+    // for surface integrals on the neighbor boundary
+    const std::vector<std::vector<Real> >& phi_neighbor_face = fe_neighbor_face->get_phi();
+    const std::vector<std::vector<RealGradient> >& dphi_neighbor_face = fe_neighbor_face->get_dphi();
+
+    // local matrices
+    DenseMatrix<Number> Ke;
+    DenseMatrix<Number> Kne;
+    DenseMatrix<Number> Ken;
+    DenseMatrix<Number> Kee;
+    DenseMatrix<Number> Knn;
```

Comment:
Could we use more descriptive variable names?

Reply:
I can try :-).